onnxruntime/core/framework/onnxruntime_typeinfo.cc

```diff
@@ -111,7 +111,8 @@ OrtStatus* OrtTypeInfo::FromDataTypeImpl(const ONNX_NAMESPACE::TypeProto* input,
         auto& t = s.dim(i);
         shape_data[i] = t.has_dim_value() ? t.dim_value() : -1;
       }
-      st = GetTensorShapeAndType(reinterpret_cast<const TensorShape*>(&shape_data), type, &info);
```

Comment:
The original code is good. Casting is a no-op at runtime.

Reply:
The original code is tricky. It's error-prone.

---
In reply to: [303077037](https://github.com/microsoft/onnxruntime/pull/1392#discussion_r303077037) [](ancestors = 303077037)

Reply:
But that is the design.

Reply:
See: https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/framework/tensor_shape.h#L135

This function has been there for a long long time. It was added by a WinML guy, and it is widely used in WinML. We must guarantee it always works.

Reply:
> TensorShape is std::vector. They have the same memory layout. You can reinterpret_cast one of them to the other.  So, there is no real use case for a move.  You can still do it, it's cheap, but it's not the best way.

This is the implementation detail of type `TensorShape`. Since `TensorShape` is private inherited from `std::vector`, this implementation detail is hidden from those who use this class. Anyone who use `TensorShape` does not need to know, and should not assume this detail.

Reply:
@fs-eire  In general, you're right. But in this project, for this class('TensorShape'), we can treat it special. Because it's so important, every dev should know it is just an alias of std::vector<int64_t>, like a typedef. 
For example, you may found this class is strange. In general, every inherited class must has a virtual destructor, otherwise the object may not get deleted correctly . But this class doesn't have it, and std::vector<int64_t> doesn't have it too.  It's kind of violating the basic law, but it doesn't mean the code is bad.  It designed as that, so every user should know the contract and obey it. Otherwise what's the meaning of inheriting from std::vector? We can easily implement this class without bothering std::vector. We do the inheritance only because we want to do the casting.

Reply:
The virtual destructor applies for public inheritance only; so it's not violating any law here. If you're publicly inheriting without a virtual destructor, it is indeed bad code; there's no exception to that. Private inheritance means "is implemented in terms of". This means the vector is purely an implementation detail of the TensorShape class. The reason for this inheritance here is not just to do casting.